### PR TITLE
fix: CNF with mutually exclusive atoms reduction

### LINKED ISSUE #57400: Wrong query result with convert_query_to_cnf + optimize_using_constraints
Comes from 02725_cnf_large_check.sql:

Setup
```sql
CREATE TABLE 02725_cnf (c0 UInt8, c1 UInt8, c2 UInt8, c3 UInt8, c4 UInt8, c5 UInt8, c6 UInt8, c7 UInt8, c8 UInt8, c9 UInt8) ENGINE = Memory;

INSERT INTO 02725_cnf VALUES (0, 0, 0, 0, 0, 0, 0, 0, 0, 0), (0, 0, 0, 0, 0, 0, 0, 0, 0, 1), (0, 0, 0, 0, 0, 0, 0, 0, 1, 0), (0, 0, 0, 0, 0, 0, 0, 0, 1, 1), (0, 0, 0, 0, 0, 0, 0, 1, 0, 0), (0, 0, 0, 0, 0, 0, 0, 1, 0, 1), (0, 0, 0, 0, 0, 0, 0, 1, 1, 0), (0, 0, 0, 0, 0, 0, 0, 1, 1, 1);
```

Default settings:
```
SELECT count()
FROM `02725_cnf`
WHERE ((NOT c2) AND c2 AND (NOT c1)) OR ((NOT c2) AND c3 AND (NOT c5)) OR ((NOT c7) AND (NOT c8)) OR (c9 AND c6 AND c8 AND (NOT c8) AND (NOT c7))

Query id: b2f5bc62-1eec-40d4-bf48-720233d7c611

┌─count()─┐
│       2 │
└─────────┘
```

With convert_query_to_cnf and optimize_using_constraints:

```
SELECT count()
FROM `02725_cnf`
WHERE ((NOT c2) AND c2 AND (NOT c1)) OR ((NOT c2) AND c3 AND (NOT c5)) OR ((NOT c7) AND (NOT c8)) OR (c9 AND c6 AND c8 AND (NOT c8) AND (NOT c7))
SETTINGS convert_query_to_cnf = 1, optimize_using_constraints = 1

Query id: b1fc0d50-e581-4211-aa35-4ef0f3bf8903

┌─count()─┐
│       8 │
└─────────┘

1 row in set. Elapsed: 0.008 sec. 
```

Enabling only one of them does not trigger the problem. Fails both with and without the analyzer.

### PR DESCRIPTION
**Pull request by korowa**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed CNF simplification, in case any OR group contains mutually exclusive atoms.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Run these jobs only (required builds will be added automatically):
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with aarch64
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Deny these jobs:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

Closes #57400 

Currently [reduceOnceCNFStatemtns](https://github.com/ClickHouse/ClickHouse/blob/3ff1b20e05c9281a4550b181a78b6a659365d69e/src/Interpreters/TreeCNFConverter.h#L168) may produce incorrect result in case any of statements is OR group with mutually exclusive atoms (e.g. `x OR !x`). Added an additional check to prevent incorrect results + prior filtering of such groups as "always true".

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 027a6a13d76e8ab6c5bcd57b9bc7cf22ffffe072 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64256/027a6a13d76e8ab6c5bcd57b9bc7cf22ffffe072/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Interpreters/TreeCNFConverter.h [hunk 1]
```diff
@@ -175,10 +181,19 @@ TAndGroup reduceOnceCNFStatements(const TAndGroup & groups)
         bool inserted = false;
         for (const auto & atom : group)
         {
-            copy.erase(atom);
             using AtomType = std::decay_t<decltype(atom)>;
             AtomType negative_atom(atom);
             negative_atom.negative = !atom.negative;
+
+            // Sikpping erase-insert for mutually exclusive atoms within
+            // signle group, since it won't insert negative atom, which
```
Comment (antonio2368) [inline src/Interpreters/TreeCNFConverter.h:189]:
```suggestion
            // single group, since it won't insert negative atom, which
```

**Review by antonio2368 — APPROVED** [review]

File: tests/queries/0_stateless/03161_cnf_reduction.sql
[the remark this answers is not in the document]
Comment (korowa) [reply tests/queries/0_stateless/03161_cnf_reduction.sql:1]: Done, moved from 02725 to 03161

**Comment by korowa** [comment]
Stress tests failure looks unrelated -- same reason as  #64053

**Comment by korowa** [comment]
@antonio2368 is it ok if I merge->rebase? Auto merge might not be triggered due to the failures, and there were some modifications for upgrade check last week, probably fixing an issue with the job in this PR

**Comment by antonio2368** [comment]
@korowa I tried to restart our private workflow which is blocking the merge. If it doesn't work you can merge master into this to restart the CI

**Comment by korowa** [comment]
@antonio2368 got it, thank you.

Will do, then, if it still will be open.

**Pull request merged**